File: generate_parentheses.py
'''
Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.

For example, given n = 3, a solution set is:

"((()))", "(()())", "(())()", "()(())", "()()()"
'''

'''
DFS. number of left parenthesis should always not exceed number of right parentheiss. Time complexity is O(2^n)
'''


class Solution:
    # @param an integer
    # @return a list of string
    def generateParenthesis(self, n):
        res = []
        self.generateparenthesisHelper(res, 0, 0, '', n)
        return res
        
    def generateparenthesisHelper(self, res, left, right, current_str, n):
        if left == right == n:
            res.append(current_str)
        elif left < right:
            return 
        else:
            if left < n:
                self.generateparenthesisHelper(res, left + 1, right, current_str + '(', n)
            self.generateparenthesisHelper(res, left, right + 1, current_str + ')', n)
                  
if __name__ == '__main__':
    s = Solution()
    print s.generateParenthesis(0)
    print s.generateParenthesis(1)
    print s.generateParenthesis(2)
    print s.generateParenthesis(3)

File: c++/generate_parentheses.cpp
/*
Given n pairs of parentheses, write a function to generate all combinations of well-formed parentheses.

For example, given n = 3, a solution set is:

[
  "((()))",
  "(()())",
  "(())()",
  "()(())",
  "()()()"
]
*/

/*
 *
*/

#include "helper.h"

class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        helper(res, n, "", 0);
        return res;
    }

    void helper(vector<string> &res, int n, const string &cur, int leftNum) {
        if (cur.length() == n * 2) {
            res.push_back(cur);
        }
        if (leftNum < n) {
            helper(res, n, cur + '(', leftNum + 1);
        }
        if (cur.length() - leftNum < n && leftNum > cur.length() - leftNum) {
            helper(res, n, cur + ')', leftNum);
        }
    }
};

int main() {
    Solution s;
    cout << s.generateParenthesis(1) << endl;
    cout << s.generateParenthesis(2) << endl;
    cout << s.generateParenthesis(3) << endl;
    return 0;
}
